parsing/tools/src/main/java/com/apache/calcite/parsing/tools/FindDialectParsingErrors.java

```diff
@@ -0,0 +1,427 @@
+        new ArrayList<>(errors.entrySet());
+    sortedEntries.sort(Comparator.comparing(e -> -e.getValue().count));
+    for (Map.Entry<String, ErrorType> e : sortedEntries) {
+      writeErrorGroupJsonObject(writer, e);
+    }
+    writer.endArray();
+    writer.endObject();
+    writer.close();
+  }
+
+  /**
+   * Writes a single JSON object containing the values inside the
+   * ErrorCountValue object for the provided entry.
+   *
+   * @param writer The JsonWriter setup for the output path
+   * @param entry The entry to write as a JSON Object
+   * @throws IOException If it fails to write to the output file
+   */
+  private void writeErrorGroupJsonObject(JsonWriter writer,
+      Map.Entry<String, ErrorType> entry) throws IOException {
+    writer.beginObject();
+    writer.name("errorMessageType").value(entry.getKey());
+    writer.name("count").value(entry.getValue().count);
+    writer.name("sampleQueries");
+    writer.beginArray();
+    for (String query : entry.getValue().sampleQueries) {
+      writer.value(query);
+    }
+    writer.endArray();
+    writer.name("fullError").value(entry.getValue().fullError);
+    writer.endObject();
+  }
+
+  /**
+   * Parses the error message and returns which error type it is. If the error
+   * message contains "Encountered", it will use the tokens occurring after
+   * "Encountered" as the error type. Otherwise, it will use one of the custom
+   * defined error messages inside CalciteResource.
+   *
+   * @param message The error message to process
+   * @return The error message type
+   */
+  private String processErrorMessage(String message) {
+    String encounteredToken = "Encountered \"";
+    int start = message.indexOf(encounteredToken);
+    if (start != -1) {
+      int end = message.indexOf("\"", start + encounteredToken
+          .length());
+      return message.substring(start, end + 1);
+    }
+    for (MessageFormat errorFormat : errorFormats) {
+      try {
+        errorFormat.parse(message);
+        return errorFormat.toPattern();
+      } catch (ParseException ignored) { }
+    }
+    return message;
+  }
+
+  /**
+   * Sanitizes the provided query. It will remove a semicolon from the end and
+   * replace some unicode characters that cannot be parsed by Calcite with an
+   * ASCII equivalent character.
+   *
+   * @param query The query to sanitize
+   * @return The sanitized query
+   */
+  private String sanitize(String query) {
+    HashMap<Character, Character> replaceChars = new HashMap<>();
+    replaceChars.put('\u00A0', ' ');
+    replaceChars.put('\u2013', '-');
+    replaceChars.put('\u2018', '\'');
+    replaceChars.put('\u2019', '\'');
+    replaceChars.put('\u201c', '\"');
+    replaceChars.put('\u201d', '\"');
+    replaceChars.put('\u2028', ' ');
+    for (Map.Entry<Character, Character> entry : replaceChars.entrySet()) {
+      query = query.replace(entry.getKey(), entry.getValue());
+    }
+    query = query.trim();
+    if (query.endsWith(";")) {
```

Comment:
will whitespace after a semicolon be a possible in the provided queries?

Reply:
Yeah some of the CSV files I was given had whitespace after it, but the trim() call above it should remove whitespace on both sides of the query.

Reply:
oh my bad i didn't see that line